packages/examples/cvat/exchange-oracle/src/handlers/completed_escrows.py

```diff
@@ -30,382 +27,259 @@
+
+        for job in jobs_batch:
+            job_annotations_file = _download_with_retries(
+                logger,
+                download_callback=partial(
+                    cvat_api.get_job_annotations,
+                    job.cvat_id,
+                    format_name=annotation_format,
+                ),
+                retry_callback=partial(
+                    cvat_api.request_job_annotations,
+                    job.cvat_id,
+                    format_name=annotation_format,
+                ),
+            )
 
-    def _process_skeletons_from_boxes_escrows(self):
-        logger = self.logger
+            job_assignment = job.latest_assignment
+            job_annotations[job.cvat_id] = FileDescriptor(
+                filename="project_{}-task_{}-job_{}-user_{}-assignment_{}.zip".format(
+                    job.cvat_project_id,
+                    job.cvat_task_id,
+                    job.cvat_id,
+                    job_assignment.user.cvat_id,
+                    job_assignment.id,
+                ),
+                file=job_annotations_file,
+            )
+    return job_annotations
+
+
+def _handle_escrow_validation(
+    logger: logging.Logger,
+    escrow_address: str,
+    chain_id: int,
+) -> bool:
+    try:
+        validate_escrow(chain_id, escrow_address)
+    except Exception as e:
+        logger.exception(f"Failed to handle completed projects for escrow {escrow_address}: {e}")
+        return False
+
+    # Need to work in separate transactions for each escrow, as a failing DB call
+    # (e.g. a failed lock attempt) will abort the transaction. A nested transaction
+    # can also be used for handling this.
+    with SessionLocal.begin() as session:
+        try:
+            escrow_projects = cvat_service.get_projects_by_escrow_address(
+                session, escrow_address, limit=None, for_update=ForUpdateParams(nowait=True)
+            )
+        except sa_errors.OperationalError as ex:
+            if isinstance(ex.orig, db_errors.LockNotAvailable):
+                return False
+            raise
+
+        _export_escrow_annotations(logger, chain_id, escrow_address, escrow_projects, session)
+
+    return True
 
-        # Here we can have several projects per escrow, so the handling is done in project groups
-        # Find potentially finished escrows first
+
+def handle_escrows_validations(logger: logging.Logger) -> None:
+    with SessionLocal.begin() as session:
+        escrow_validations = cvat_service.get_escrows_for_validation(
+            session,
+            limit=CronConfig.track_completed_escrows_chunk_size,
+        )
+
+    unhandled_validations = set(validation_id for validation_id, _, __ in escrow_validations)
+    handled_validations = set()
+    try:
+        for validation_id, escrow_address, chain_id in escrow_validations:
+            if _handle_escrow_validation(logger, escrow_address, chain_id):
+                handled_validations.add(validation_id)
+    finally:
```

Comment:
Shouldn't all of this be within a single transaction with select for update object locking?

Reply:
I don't think having long connections to the database is a good idea, let alone holding a lock.

Besides, we don't really need it here — `get_escrows_for_validation` updates the status of validations to `in-progress`. Though it was not obvious from its name, so I've changed the function name to `prepare_escrows_for_validation`.

Reply:
So if I understand correctly, if we get some escrows for handling and then the function fails for some reason (e.g. the exchange oracle is shut down), these escrows will never be processed? Probably, we need something different. 

> I don't think having long connections to the database is a good idea, let alone holding a lock.

Do you know about any problems with doing so?

Reply:
As a side note.

I think we would benefit here from adopting jobs, queues and workers approach. This would abstract away retries logic and will allow us to focus just on task implementation at hand.

State trackers then won't be doing any work and will just schedule jobs.

Reply:
Overall looks good.

> I think we would benefit here from adopting jobs, queues and workers approach.

Do you have some specific solution in mind? Anyway, it doesn't seem to be a problem now.